src/snooping.jl

```diff
@@ -1,94 +1,156 @@
+                # we need to wrap into try catch, since some anonymous symbols won't
+                # be found... there is also still a high probability, that some modules
+                # aren't defined
+                line_idx += 1
+                println(io, "try;", line, "; catch e; @warn \"couldn't precompile statement $line_idx\" exception = e; end")
             end
         end
-        exit()
-    end)
-    wait(process)
-    println("done.")
-    nothing
+    end
+    rm(tmp_file, force = true)
+    outputfile
 end
 
-function snoop(path, compilationfile, csv, subst, blacklist)
-    csv = String(abspath(csv))
-    execpath = String(abspath(path))
-    snoop_vanilla(String[], csv, :(include($execpath)), dirname(execpath))
-    data = SnoopCompile.read(csv)
-    #pc = SnoopCompile.parcel(reverse!(data[2]))
-    pc = SnoopCompile.format_userimg(reverse!(data[2]), subst=subst, blacklist=blacklist)
-    pc = map(x -> "try; $x; catch ex; @warn \"\"\"skipping line: $(repr(x)).\"\"\" exception=ex; end\n", pc)
-    SnoopCompile.write(compilationfile, pc)
-    nothing
-end
 
 """
     snoop_userimg(userimg, packages::Tuple{String, String}...)
 
     Traces all function calls in packages and writes out `precompile` statements into the file `userimg`
 """
-function snoop_userimg(userimg, packages::Tuple{String, String}...)
+function snoop_userimg(userimg, packages::Tuple{String, String}...; additional_packages = Symbol[])
     snooped_precompiles = map(packages) do package_snoopfile
         package, snoopfile = package_snoopfile
+        module_file = ""
         abs_package_path = if ispath(package)
-            normpath(abspath(package))
+            path = normpath(abspath(package))
+            module_file = joinpath(path, "src", basename(path) * ".jl")
+            path
         else
-            normpath(Base.find_package(package), "..", "..")
+            module_file = Base.find_package(package)
+            normpath(module_file, "..", "..")
         end
+        module_name = Symbol(splitext(basename(module_file))[1])
         file2snoop = normpath(abspath(joinpath(abs_package_path, snoopfile)))
         package = package_folder(get_root_dir(abs_package_path))
         isdir(package) || mkpath(package)
         precompile_file = joinpath(package, "precompile.jl")
-        snoop(file2snoop, precompile_file, joinpath(package, "snooped.csv"),
-            Vector{Pair{String, String}}(), String["Main"])
+        snoop(module_name, file2snoop, precompile_file; additional_packages = additional_packages)
         return precompile_file
     end
     # merge all of the temporary files into a single output
     open(userimg, "w") do output
-        println(output, """
-            # Prevent this from being put into the Main namespace
-            Core.eval(Module(), quote
-            """)
+        # Prevent this from being put into the Main namespace
+        println(output, "module CompilationModule")
```

Comment:
Yes, this is still WIP... eval was segfaulting with 19.000 statements and I couldn't figure out why, so here we are ;)

Reply:
How about generating two files, one with only the precompile statements and the other with
```
Core.eval(Module(), :(include("the_other_file.jl"))
```
or just calling that line directly instead of including a file?